pay-api/tests/conftest.py

```diff
@@ -67,160 +69,63 @@ def client_ctx(app):
-                print('DROP VIEW public.%s ' % seq)
-            except Exception as err:  # NOQA pylint: disable=broad-except
-                print(f'Error: {err}')
-        sess.commit()
-
-        # Clear out any existing tables
-        metadata = MetaData(_db.engine)
-        metadata.reflect()
-        for table in metadata.tables.values():
-            for fk in table.foreign_keys:  # pylint: disable=invalid-name
-                _db.engine.execute(DropConstraint(fk.constraint))
-        metadata.drop_all()
-        _db.drop_all()
-
-        sequence_sql = """SELECT sequence_name FROM information_schema.sequences
-                          WHERE sequence_schema='public'
-                       """
-
-        sess = _db.session()
-        for seq in [name for (name,) in sess.execute(text(sequence_sql))]:
-            try:
-                sess.execute(text('DROP SEQUENCE public.%s ;' % seq))
-                print('DROP SEQUENCE public.%s ' % seq)
-            except Exception as err:  # NOQA pylint: disable=broad-except
-                print(f'Error: {err}')
-        sess.commit()
-
-        # ############################################
-        # There are 2 approaches, an empty database, or the same one that the app will use
-        #     create the tables
-        #     _db.create_all()
-        # or
-        # Use Alembic to load all of the DB revisions including supporting lookup data
-        # This is the path we'll use in legal_api!!
-
-        # even though this isn't referenced directly, it sets up the internal configs that upgrade needs
+        if database_exists(_db.engine.url):
+            drop_database(_db.engine.url)
+        create_database(_db.engine.url)
         Migrate(app, _db)
         upgrade()
-
         return _db
 
 
 @pytest.fixture(scope='function', autouse=True)
-def session(app, db):  # pylint: disable=redefined-outer-name, invalid-name
+def session(db, app):  # pylint: disable=redefined-outer-name, invalid-name
     """Return a function-scoped session."""
     with app.app_context():
-        conn = db.engine.connect()
-        txn = conn.begin()
+        with db.engine.connect() as conn:
+            transaction = conn.begin()
+            sess = db._make_scoped_session(dict(bind=conn, autoflush=False))  # pylint: disable=protected-access
+            # Establish SAVEPOINT (http://docs.sqlalchemy.org/en/latest/orm/session_transaction.html#using-savepoint)
+            nested = sess.begin_nested()
+
+            @event.listens_for(sess, 'after_transaction_end')
+            def restart_savepoint(sess2, trans):  # pylint: disable=unused-variable
+                nonlocal nested
+                if trans.nested:
+                    # Handle where test DOESN'T session.commit()
+                    sess2.expire_all()
+                    nested = sess.begin_nested()
+                    # When using a SAVEPOINT via the Session.begin_nested() or Connection.begin_nested() methods,
```

Comment:
Self explainatory.. if you commit or rollback on a session or a connection, it rolls back the outer object.. which is why we needed to patch the commit and rollbacks in nested operations.

Reply:
This was a pain to get going, alot of unit tests failing strangely - there might be an easier way to do this.